src/runtime/refiner.ts

```diff
@@ -8,142 +8,699 @@
+    }
+
+    toString(): string {
+        return '[' + this.expression.toString() + ']';
+    }
+
+    validateData(data: Dictionary<ExpressionPrimitives>): boolean {
+        const res = this.expression.applyOperator(data);
+        if (typeof res !== 'boolean') {
+            throw new Error('Refinement expression evaluated to a non-boolean type.\n');
         }
-        return expr.value.toString();
+        return res;
+    }
+}
+
+abstract class RefinementExpression {
+    type: 'Boolean' | 'Number' | 'Text';
+
+    constructor(readonly kind: 'binary-expression' | 'unary-expression' | 'field-name' | 'number' | 'boolean') {}
+
+    static fromAst(expr: RefinementExpressionNode, typeData: Dictionary<ExpressionPrimitives>): RefinementExpression {
+        switch (expr.kind) {
+            case 'binary-expression-node': return new BinaryExpression(expr, typeData);
+            case 'unary-expression-node': return new UnaryExpression(expr, typeData);
+            case 'field-name-node': return new FieldNamePrimitive(expr, typeData);
+            case 'number-node': return new NumberPrimitive(expr);
+            case 'boolean-node': return new BooleanPrimitive(expr);
+            default: throw new Error('Unknown node type.');
+        }
+    }
+
+    normalise(): RefinementExpression {
+        return this;
+    }
+
+    abstract toString();
+
+    abstract applyOperator(data: Dictionary<ExpressionPrimitives>);
+}
+
+export class BinaryExpression extends RefinementExpression {
+    type: 'Number' | 'Boolean';
+    leftExpr: RefinementExpression;
+    rightExpr: RefinementExpression;
+    operator: RefinementOperator;
+
+    constructor(expression: BinaryExpressionNode, typeData: Dictionary<ExpressionPrimitives>) {
+        super('binary-expression');
+        this.leftExpr = RefinementExpression.fromAst(expression.leftExpr, typeData);
+        this.rightExpr = RefinementExpression.fromAst(expression.rightExpr, typeData);
+        this.operator = new RefinementOperator(expression.operator);
+        this.operator.assertOperandCompatibility([this.leftExpr.type, this.rightExpr.type]);
+        this.type = this.operator.evalType();
+    }
+
+    toString(): string {
+        return '(' + this.leftExpr.toString() + ' ' + this.operator.op + ' ' + this.rightExpr.toString() + ')';
+    }
+
+    applyOperator(data: Dictionary<ExpressionPrimitives>): ExpressionPrimitives {
+        const left = this.leftExpr.applyOperator(data);
+        const right = this.rightExpr.applyOperator(data);
+        return this.operator.eval([left, right]);
+    }
+
+    swapChildren() {
+        const temp = this.rightExpr;
+        this.rightExpr = this.leftExpr;
+        this.leftExpr = temp;
+        switch (this.operator.op) {
+            case '<': this.operator.updateOp('>'); break;
+            case '>': this.operator.updateOp('<'); break;
+            case '<=': this.operator.updateOp('>='); break;
+            case '>=': this.operator.updateOp('<='); break;
```

Comment:
There's a lot of bare string values used quite widely here ('and', '<=', 'neg', 'binary-expression', etc). It's usually a good idea to make those string-valued enums. It avoids typo errors, aids TS's type checking and enables concise type declarations:

```
enum ExprKind {
  Binary  = 'binary-expression',
  Unary   = 'unary-expression',
  Field   = 'field-name',
  Number  = 'number',
  Boolean = 'boolean',
}

enum Op {
  And = 'and',
  Or  = 'or',
  LT  = '<',
  GT  = '>',
  LTE = '<=',
  GTE = '>=',
  Add = '+',
  Sub = '-',
  Mul = '*',
  Div = '/',
  Not = 'not',
  Neg = 'neg',
}

abstract class RefinementExpression {
 constructor(readonly kind: ExprKind) {}
 ...
}
```